arm_isa.cpp

```diff
@@ -1377,7 +1377,7 @@ inline void LDRH(arm_isa* ref, int rd, int rn,
     printf("Unpredictable LDRH instruction result (Address is not Halfword Aligned)\n");
     return;
   }
-  value = MEM.read(ref->ls_address.entire);
+  value = MEM.read_half(ref->ls_address.entire);
   value &= 0xFFFF; /* Zero extends halfword value 
 		      BUG: Model must be little endian in order to the code work  */
```

Comment:
In the future it would be nice to have a test system that allowed a bug fix commit to bring together a testcase file showing the exact bug it fixes.

It's funny because "read_half" shouldn't need a mask. If it needs a mask, it means it is sign-extending the value loaded from memory. Ideally it should just return a 16-bit unsigned number and let the compiler perform the sign-extension only if you cast it to a signed 16-bit number. Well, this is ultimately a judgement call, but the way I see it is that sign-extension can be pretty harmful for low-level simulators because it can generate unexpected values when converting from smaller to larger types, and that we should avoid sign-extension as much as possible, except when explicitly asked via casts to signed types.

Another thing I noticed is that this cache simulator probably doesn't accept misaligned memory accesses, right? Because it was crashing with arm loading a 32-bit number aligned to 16-bit boundary. This is something to watch for, because I think more models may be liberal with respect to memory access (I mean, they don't care if the memory accesses are aligned or not).